sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,139 @@
+
+import contextlib
+import json
+import os
+import shutil
+from sos.collector.transports import RemoteTransport
+from sos.collector.exceptions import ConnectionException
+from sos.utilities import (is_executable,
+                           sos_get_command_output)
+
+
+class SaltStackMasterTransport(RemoteTransport):
+    """
+    A transport for collect that leverages SaltStack's Master Pub/Sub
+    functionality to send commands to minions.
+
+    This transport will by default assume the use cmd.shell module to
+    execute commands on the minions.
+    """
+
+    name = 'saltstack'
+
+    def _convert_output_json(self, json_output):
+        return list(json.loads(json_output).values())[0]
+
+    def run_command(
+            self, cmd, timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Run a command on the remote host using SaltStack Master.
+        If the output is json, convert it to a string.
+        """
+        ret = super(SaltStackMasterTransport, self).run_command(
+            cmd, timeout, need_root, env, get_pty)
+        with contextlib.suppress(Exception):
+            ret['output'] = self._convert_output_json(ret['output'])
+        return ret
+
+    def _salt_retrieve_file(
+            self, node, fname, dest,
+            timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Execute cp.push on the remote host using SaltStack Master
+        """
+        cmd = f"salt {node} cp.push {fname}"
+        res = sos_get_command_output(cmd)
+        if res['status'] == 0:
+            cachedir = f"/var/cache/salt/master/minions/{self.address}/files"
+            cachedir_file = os.path.join(cachedir, fname.lstrip('/'))
+            shutil.move(cachedir_file, dest)
+            return True
+        return False
+
+    @property
+    def connected(self):
+        """Check if the remote host is responding using SaltStack Master."""
+        up = self.run_command("echo Connected", timeout=10)
+        return up['status'] == 0
+
+    def _check_for_saltstack(self, password=None):
+        """Checks to see if the local system supported SaltStack Master.
+
+        This check relies on feedback from the salt binary. The command being
+        run should always generate stderr output, but depending on what that
+        output reads we can determine if SaltStack Master is supported or not.
+
+        For our purposes, a host that does not support SaltStack Master is not
+        able to run sos-collector.
+
+        Returns
+            True if SaltStack Master is supported, else raise Exception
+        """
+
+        def _check_for_salt_master(command):
+            """Check for various salt master binaries on the local system."""
+            res = sos_get_command_output(command)
+            return res['status'] == 0
+
+        cmd = 'salt-run manage.status'
+        res = _check_for_salt_master(cmd)
+        return res
```

Comment:
Until there are plans for possible re-usage of the `_check_for_salt_master` routine, why not calling directly:

```
res = sos_get_command_output('salt-run manage.status')
return res['status'] == 0
```

?

Reply:
Now that salt binary is handled by **is_executable()** instead, no reason. I'll move this to **sos_get_command_output()** and leave any alternative logic for a cluster implementation.